solver/progress.go

```diff
@@ -111,44 +116,56 @@ func NewProgress(ctx context.Context, opts ...ProgressOption) (Progress, error)
 		}
 	}
 
-	// Not using shared context to not disrupt display on errors, and allow
-	// graceful exit and report error.
-	pctx, cancel := context.WithCancel(context.Background())
-
-	var pw *progress.Printer
-
+	var mode string
 	switch info.LogOutput {
 	case LogOutputTTY:
-		pw = progress.NewPrinter(pctx, info.Console, "tty")
+		mode = "tty"
 	case LogOutputPlain:
-		pw = progress.NewPrinter(pctx, info.Console, "plain")
+		mode = "plain"
 	default:
-		cancel()
 		return nil, errors.Errorf("unknown log output %q", info.LogOutput)
 	}
 
-	g, ctx := errgroup.WithContext(ctx)
-
-	done := make(chan struct{})
-	g.Go(func() error {
-		defer cancel()
-		<-done
-		return pw.Wait()
+	spp, wait, release := newSyncProgressPrinter(info.Console, mode)
+	p := &progressUI{
+		origCtx: ctx,
+		spp:     spp,
+		mw:      NewMultiWriter(spp),
+		done:    make(chan struct{}),
+	}
+	p.g, p.ctx = errgroup.WithContext(p.origCtx)
+	p.g.Go(func() error {
+		defer release()
+		<-p.done
+		return wait()
 	})
+	return p, nil
+}
 
-	return &progressUI{
-		mw:   NewMultiWriter(pw),
-		ctx:  ctx,
-		g:    g,
-		done: done,
-	}, nil
+func (p *progressUI) Sync() error {
+	p.Release()
+	err := p.Wait()
+	if err != nil {
+		return err
+	}
+	wait, release := p.spp.reset()
+	p.g, p.ctx = errgroup.WithContext(p.origCtx)
+	p.done = make(chan struct{})
+	p.g.Go(func() error {
+		defer release()
+		<-p.done
+		return wait()
+	})
```

Comment:
Curious why we need a goroutine that just sits around and waits for a call to `Release`. Couldn't we just have `p.Wait()` block on `p.done` and call `wait`?

Reply:
Yeah, this was mostly old stuff that I just tweaked a bit.  I have refactored to get rid of this gratuitous use of the errGroup.   I also ended up removing `Go` from the progress interface.  That usage was also gratuitous and prevented me from calling. Sync before exec'ing into the error container.